[TopKec] iP

TopKec opened the pull request:
```
 _____ _                       
/  ___| |                      
\ `--.| |_ _____   _____ _ __  
 `--. \ __/ _ \ \ / / _ \ '_ \ 
/\__/ / ||  __/\ V /  __/ | | |
\____/ \__\___| \_/ \___|_| |_|
                                                          
```
Need advice? Making a list? Steven's the bot!

### Features:
- Creating lists of actions to make
- Saving said actions to be read again later
- Marking, deleting and so forth of tasks

### Useful commands in the bot
- [x] `help` whenever you're not sure what to do
- [x] `list` to view what you have
- [x] `bye` to exit the program
- [ ] and so much more! :smiley:

# Sounds great! How do I download it?
1. Visit [this](https://github.com/TopKec/ip)
2. Compile it on your machine.
3. Let it run! **Its as simple as that!**

> Steven's advice: Don't wait! Install now!

~~Please understand that Steven is not indeed sentient~~

*Text at top of this readme has been provided by [parorjk.com](https://patorjk.com/software/taag/)*

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

TopKec committed:
feat: Removed Duke.java and added Steven.java

TopKec committed:
feat: Added support for user input (echo) and exiting the program

TopKec committed:
feat: Removed echo feature, added functionality to add things to a list, and note the items that exist in the list

TopKec committed:
Revert "feat: Removed echo feature, added functionality to add things to a list, and note the items that exist in the list"

This reverts commit 1c125e30c3dd330f9672cf0471d9eb83ba973f98.

TopKec committed:
feat: Removed echo feature, added functionality to add things to a list, and note the items that exist in the list

TopKec committed:
fix: commited Task.java, which would allow Steven to function

TopKec committed:
feat: added functionality to marking and unmarking of items in the list

TopKec committed:
feat: Added support for adding different types of events to be added to the list, namely: todo, deadline and event

TopKec committed:
fix: fixed a bug where Steven would not be able to accept items with a space in their names, resulting in truncating to the first word in the name of the item added.

TopKec committed:
fix: Fixed a bug which resulted in errors when attempting to run Steven in an automated manner

TopKec committed:
test: Updated files to allow for automated running of Steven

TopKec committed:
test: updated gitignore

TopKec committed:
feat: Provided functionality to assist the user if inputting an invalid input. Also created a list of commands to assist user if one doesn't know what commands to use.

TopKec committed:
feat: added functionality for deleting of items in the list

TopKec committed:
Add support for saving data

TopKec committed:
Update .gitignore file

TopKec committed:
Update .gitignore file

TopKec committed:
Modify code to make user input a class instead of a raw string

TopKec committed:
Mandate dates for "Event" and "Deadline" to be ISO-8601-compliant

TopKec committed:
Merge branch 'branch-level-8'

TopKec committed:
Convert program to more OO solution

TopKec committed:
Move files into appropriate packages

TopKec committed:
Fix bug where Steven can not exit

TopKec committed:
Merge remote-tracking branch 'origin/add-gradle-support'

TopKec committed:
Add support for Gradle

TopKec committed:
Add Support for JUnit tests

TopKec committed:
Create .jar file

TopKec committed:
Adjust code to be compliant with coding standard

TopKec committed:
Add Javadoc for files

TopKec committed:
Add support for find operation

TopKec committed:
Merge branch 'branch-level-9'

TopKec committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Parser.java

TopKec committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Parser.java
#	src/main/java/Steven.java
#	src/main/java/Storage.java
#	src/main/java/TaskList.java
#	src/main/java/tasks/Deadline.java
#	src/main/java/tasks/Event.java
#	src/main/java/tasks/Todo.java

TopKec committed:
Organise more files into packages

laksherthanlife commented on docs/README.md:
I like how you have begun the process of updating your README file even though it is not an expectation

laksherthanlife commented on src/main/java/Steven.java:
```diff
@@ -0,0 +1,62 @@
+public class Steven {
+
+    private static final String line = "========\n";
+    private static final String bootMsg = ("This is Steven!\nHow can I advise?\n");
```
Do you think it would be more aligned with the style guide to name the variable messageWhenBooting? I feel that would be more noun like

laksherthanlife commented on src/main/java/Steven.java:
```diff
@@ -0,0 +1,62 @@
+            Scanner input = new Scanner(System.in);
+            while (input.hasNextLine()) {
+                UserInput command = new UserInput(input.nextLine());
+                exit = new Parser(command, this.tasks).processInput();
```
Do you think it would be better to name this variable exitStatement?

laksherthanlife commented on src/main/java/exceptions/ExcessiveArgException.java:
```diff
@@ -0,0 +1,8 @@
+ * Exception class for Steven
+ * Used for when there are excessive arguments for a given command
+ */
+public class ExcessiveArgException extends Exception{
```
Can we consider having a space before the brace to better follow the coding standard guidelines?

laksherthanlife commented on src/main/java/mainFiles/Parser.java:
```diff
@@ -0,0 +1,260 @@
+                        + "find (item) - item is the name of the task, or part of the task");
+            }
+            break;
+        case "help":
```
I like how you have a help case!

laksherthanlife commented on src/main/java/mainFiles/UserInput.java:
```diff
@@ -0,0 +1,81 @@
+    private final String name;
+    private final String arg1;
+    private final String arg2;
+    private final String arg3;
```
do you think it would be clearer to name these instances fromDate, byDate, and toDate respectively?

TopKec replied:
Ah, in this case I've left these as more generic arguments because they're meant to support other things, so for example, in the case of `mark`ing, it accepts the input as `mark x`, but x is an integer, not necessarily a date.

laksherthanlife commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate start;
```
do you think you should name these startDate and toDate respectively?

TopKec replied:
I do think so! I'll be making the adjustments promptly. Also, as a general note, I'll likely be implementing most of the changes you listed, just didn't want to comment on every single comment, haha. Thanks very much for the review!

laksherthanlife reviewed (commented):
LGTM. Just a few nits to fix!

TopKec committed:
Update README.md

TopKec committed:
Fix code to be more aligned to coding standards

TopKec committed:
Move actions into their respective classes

TopKec committed:
Add UI Support

TopKec committed:
Fix "mark" and "unmark" commands for Steven

Steven has commands to mark (or unmark) a given task as "complete"
or "incomplete" respectively.

There is a bug which respectively caused mark and unmark to mark
given task as "incomplete" or "complete", and also prevents
tasks as being marked or unmarked when incomplete and complete
respectively

This is unintended behavior.

Bugfix should invert the above behavior

Let's
* allow marking to be done on incomplete tasks, and set them to
complete
* unmarking to be done on complete tasks, and set them to
incomplete

TopKec committed:
Fix dialog box not resizing

For Steven, the dialog box does not expand with more text, so for
example, a very long command will appear cut-off with ellipsis.

This prevents a a user from viewing the full message from Steven.

This is unintended behavior and thus should be fixed.

Let's,
* Fix the above bug so the full output from Steven is properly
observed

TopKec committed:
Update to support building of jar file

TopKec committed:
Update README.md

TopKec committed:
Improve code quality

TopKec committed:
Add assertions

TopKec committed:
Merge pull request #1 from TopKec/branch-A-Assertions

Branching Assertions and Code quality

TopKec committed:
Add support for "sort" command

TopKec committed:
Update README file

TopKec committed:
Improve GUI functionality

TopKec committed:
Merge branch 'A-BetterGui'

TopKec committed:
Fix bug where "bye" does not exit program

TopKec committed:
Add proper credit for previous bugfix

TopKec committed:
Move files into appropriate packages

TopKec committed:
Move more files into packages

TopKec committed:
Fix more bugs relating to the name "Steven" in strings

TopKec committed:
Add UI image

TopKec committed:
Update README file

TopKec committed:
Improve code quality

TopKec committed:
Fix bug with file creation

TopKec committed:
Fix issue for test files being in gitignore
